presto-native-execution/presto_cpp/main/TaskManager.cpp

```diff
@@ -256,6 +257,102 @@ void TaskManager::getDataForResultRequests(
   }
 }
 
+namespace {
+std::unordered_map<std::string, std::string> toConfigs(
+    const protocol::SessionRepresentation& session) {
+  auto configs = std::unordered_map<std::string, std::string>(
+      session.systemProperties.begin(), session.systemProperties.end());
+
+  // If there's a timeZoneKey, convert to timezone name and add to the
+  // configs. Throws if timeZoneKey can't be resolved.
+  if (session.timeZoneKey != 0) {
+    configs.emplace(
+        velox::core::QueryConfig::kSessionTimezone,
+        velox::util::getTimeZoneName(session.timeZoneKey));
+  }
+  return configs;
+}
+
+std::unordered_map<std::string, std::unordered_map<std::string, std::string>>
+toConnectorConfigs(const protocol::SessionRepresentation& session) {
+  std::unordered_map<std::string, std::unordered_map<std::string, std::string>>
+      connectorConfigs;
+  for (const auto& entry : session.catalogProperties) {
+    connectorConfigs.insert(
+        {entry.first,
+         std::unordered_map<std::string, std::string>(
+             entry.second.begin(), entry.second.end())});
+  }
+
+  return connectorConfigs;
+}
+
+/// Presto-on-Spark is expected to specify all splits at once along with
+/// no-more-splits flag. Verify that all plan nodes that require splits
+/// have received splits and no-more-splits flag. This check helps
+/// prevent hard-to-debug query hangs caused by Velox Task waiting for
+/// splits that never arrive.
+void checkSplitsForBatchTask(
+    const velox::core::PlanNodePtr& planNode,
+    const std::vector<protocol::TaskSource>& sources) {
+  std::unordered_set<velox::core::PlanNodeId> splitNodeIds;
+  velox::core::PlanNode::findFirstNode(
+      planNode.get(), [&](const velox::core::PlanNode* node) {
+        if (node->requiresSplits()) {
+          splitNodeIds.insert(node->id());
+        }
+        return false;
+      });
+
+  for (const auto& source : sources) {
+    VELOX_USER_CHECK(
+        source.noMoreSplits,
+        "Expected no-more-splits message for plan node {}",
+        source.planNodeId);
+    splitNodeIds.erase(source.planNodeId);
+  }
+
+  VELOX_USER_CHECK(
+      splitNodeIds.empty(),
+      "Expected all splits and no-more-splits message for all plan nodes: {}",
+      folly::join(", ", splitNodeIds));
+}
+} // namespace
+
+std::unique_ptr<protocol::TaskInfo> TaskManager::createOrUpdateTask(
+    const protocol::TaskId& taskId,
+    const protocol::TaskUpdateRequest& updateRequest,
+    const velox::core::PlanFragment& planFragment) {
+  const auto& session = updateRequest.session;
+
+  return createOrUpdateTask(
+      taskId,
+      planFragment,
+      updateRequest.sources,
+      updateRequest.outputIds,
+      toConfigs(session),
+      toConnectorConfigs(session));
+}
+
+std::unique_ptr<protocol::TaskInfo> TaskManager::createOrUpdateBatchTask(
```

Comment:
Will this path be called for Presto Batch too?
Or is this only called by the spark code?

Reply:
This is for Spark only.